app/controllers/administrate/application_controller.rb

```diff
@@ -106,8 +106,19 @@ def nav_link_state(resource)
     end
 
     helper_method :valid_action?
-    def valid_action?(name, resource = resource_class)
-      routes.include?([resource.to_s.underscore.pluralize, name.to_s])
+    def valid_action?(name, resource = resource_class, parent_resource = nil)
+      !!routes.detect do |controller, action, required_parts|
```

Comment:
This change from using `routes.include?` to using `routes.detect`. This brings us back to a O(n) algorithm, which is slow for apps with many routes: https://github.com/thoughtbot/administrate/pull/2182.

Reply:
Indeed. Thanks for the heads up.

I've tried to change the data structure to a hash, to keep the O(1) access of the route in the routes list and be able to store “required action parts” too.

I guess it would require more memory, but would keep the runtime access complexity similar, do you think that would be suitable?

Reply:
A hash would probably work too, yes.